src/ssh/cockpitsshrelay.c

```diff
@@ -477,35 +436,39 @@ write_tmp_knownhosts_file (CockpitSshData *data,
                            const gchar *content,
                            const gchar **problem)
 {
-  FILE *fp = NULL;
+  int fd;
+  size_t content_len = strlen (content);
+  g_autoptr(GError) error = NULL;
 
-  tmp_knownhost_file = create_knownhosts_temp ();
-  if (!tmp_knownhost_file)
+  fd = g_file_open_tmp ("known-hosts.XXXXXX", &tmp_knownhost_file, &error);
+  if (fd < 0)
     {
+      g_warning ("%s: couldn't open temporary known host file for data: %s",
+                 data->logname, error->message);
       *problem = "internal-error";
       return FALSE;
     }
+
   atexit (cleanup_knownhosts_file);
 
-  fp = fopen (tmp_knownhost_file, "a");
-  if (fp == NULL)
+  for (;;)
```

Comment:
I think maybe you planned to try multiple `write()` in case of a partial result, but you never actually implemented this (instead opting to give an error message in this case).  This loop only ever runs once, as a result.

Calling `posix_fallocate()` first is a pretty reasonable compromise here (and generally improves results on extents-based filesystems).

Reply:
The loop was just meant to retry on `EINTR`. But indeed that writes a file and may get a short write; I think this is not really worth the trouble, and I'll just move this to `g_file_set_contents()`. This has absolutely no performance requirements, and once `create_knownhosts_temp()` succeeds, we can close the fd and reopen the file.